# Extração da característica relação sinal-ruído (SNR) de dados de EEG

A relação sinal-ruído (SNR, do inglês _Signal-to-Noise Ratio_) é uma medida que compara o nível do sinal desejado com o nível do ruído de fundo. Uma SNR mais alta indica um sinal mais forte em comparação ao ruído, enquanto uma SNR mais baixa indica que o ruído é comparativamente forte.

O objetivo é utilizar um conjunto de dados com nível de atividade cerebral basal (que será considerado e utilizado como "ruído") e "removê-lo" do sinal obtido no experimento a fim de reconhecer os padrões de foco com maior clareza.

No contexto da caracterização de foco, podemos, a partir do sinal basal, classificar os sinais dos ritmos cerebrais com a presença ou não de foco, de forma que as amostras de sinais extraídas de um buffer sejam rotuladas com com a presença ou não de foco.

Esta análise será realizada com o auxílio do classificador SVM (_Support Vector Machine_). Nesse sentido, uma porcetagem das amostras são utilizadas para treino e outra para o teste (30 e 70% respectivamente).

In [16]:
import mne
import numpy as np
from scipy.signal import welch
import matplotlib.pyplot as plt
from matplotlib import rcParams

In [17]:
# Transform a string that represents a time value ("min:sec") into number and converts it to seconds
def convert_min_to_sec(time):
    minutes, seconds = map(int, time.split(":"))
    return minutes * 60 + seconds

# Transfor a string thats represents a range time "0:21 - 0:40" into index that will be used to cut data
def convert_time_range_to_index(timerange):
    start, end = map(str, timerange.split(" - "))
    new_start = convert_min_to_sec(start)
    new_end = convert_min_to_sec(end)
    
    index = []
    index.append(new_start * 250)
    index.append(new_end * 250)

    return index

In [18]:
n_channels = 8
ch_types = ['eeg'] * n_channels
sfreq = 250
ch_names = ["F3", "Fz", "F4", "C3", "Cz", "C4", "P3", "P4"]
info = mne.create_info(ch_names=ch_names, sfreq=sfreq, ch_types=ch_types)
info.set_montage("standard_1020")

<Info | 8 non-empty values
 bads: []
 ch_names: F3, Fz, F4, C3, Cz, C4, P3, P4
 chs: 8 EEG
 custom_ref_applied: False
 dig: 11 items (3 Cardinal, 8 EEG)
 highpass: 0.0 Hz
 lowpass: 125.0 Hz
 meas_date: unspecified
 nchan: 8
 projs: []
 sfreq: 250.0 Hz
>

In [19]:
ia = '../dataset-s7/IA/OpenBCI-RAW-2023-09-28_16-51-25_IA.txt'
ia_ob = np.loadtxt(ia, delimiter=',', skiprows=5, usecols=range(1, 9))

basal = '../dataset-s7/TF/OpenBCI-RAW-2023-11-07_13-17-01_TF.txt'
basal_ob = np.loadtxt(basal, delimiter=',', skiprows=5, usecols=range(1, 9))

In [20]:
# Aula IA
ia_timeranges = [
    "6:00 - 10:17",
]

# Basal (TF)
basal_timerange = [
    "3:28 - 4:28"
]

In [21]:
ia_index = []
ia_index = convert_time_range_to_index(ia_timeranges[0])

basal_index = []
basal_index = convert_time_range_to_index(basal_timerange[0])

data_cut_ia = ia_ob[ia_index[0]:ia_index[1], :]
data_cut_basal = basal_ob[basal_index[0]:basal_index[1], :]

X = {
    'ia': mne.io.RawArray(data_cut_ia.T, info),
    'basal': mne.io.RawArray(data_cut_basal.T, info),
}

Creating RawArray with float64 data, n_channels=8, n_times=64250
    Range : 0 ... 64249 =      0.000 ...   256.996 secs
Ready.
Creating RawArray with float64 data, n_channels=8, n_times=15000
    Range : 0 ... 14999 =      0.000 ...    59.996 secs
Ready.


### Filtragem dos dados

O laço a seguir aplica dois filtros nos dados. Primeiramente, atenua-se com um filtro _notch_ a frequência de 60Hz de forma a evitar a interferência da rede elétrica e, posteriormente, obtém-se a faixa de frequência desejada de 4 a 100Hz com um filtro passa faixa.

In [22]:
# TODO: filtrar corretamente os dados que são usados para SNR (basal)

for key in X:
    X[key].notch_filter(freqs=60)
    X[key].filter(l_freq=4, h_freq=100)

# X['ia'].notch_filter(freqs=60)
# X['ia'].filter(l_freq=4, h_freq=100)

Filtering raw data in 1 contiguous segment
Setting up band-stop filter from 59 - 61 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandstop filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 59.35
- Lower transition bandwidth: 0.50 Hz (-6 dB cutoff frequency: 59.10 Hz)
- Upper passband edge: 60.65 Hz
- Upper transition bandwidth: 0.50 Hz (-6 dB cutoff frequency: 60.90 Hz)
- Filter length: 1651 samples (6.604 s)

Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 4 - 1e+02 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 4.00
- Lower transition bandwidth: 2.00 Hz (-6 dB cutoff frequency: 3.00 Hz)
- Upper passband e

Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 4 - 1e+02 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 4.00
- Lower transition bandwidth: 2.00 Hz (-6 dB cutoff frequency: 3.00 Hz)
- Upper passband edge: 100.00 Hz
- Upper transition bandwidth: 25.00 Hz (-6 dB cutoff frequency: 112.50 Hz)
- Filter length: 413 samples (1.652 s)



In [ ]:
def label_data(freqs, psd, samples):
    
    # Definindo os limites das bandas de frequência (em Hz)
    theta_band = (4, 8)       # Theta: 4 - 8 Hz
    alpha_band = (8, 13)      # Alpha: 8 - 13 Hz
    beta_band = (13, 30)      # Beta: 13 - 30 Hz
    gamma_band = (30, 100)    # Gamma: 30 - 100 Hz
    
    # Calculando a média dos 8 eletrodos (transformando em um ndarray unidimensional)
    X_avg = np.average(psd, axis=0)

    # Encontrar os índices correspondentes às frequências de interesse
    theta_idxs = np.where((freqs >= theta_band[0]) & (freqs <= theta_band[1]))[0]
    alpha_idxs = np.where((freqs >= alpha_band[0]) & (freqs <= alpha_band[1]))[0]
    beta_idxs = np.where((freqs >= beta_band[0]) & (freqs <= beta_band[1]))[0]
    gamma_idxs = np.where((freqs >= gamma_band[0]) & (freqs <= gamma_band[1]))[0]

    # Calculando a potência em cada banda de frequência por meio da integração da PSD
    bands = [
        np.sum(X_avg[theta_idxs]),
        np.sum(X_avg[alpha_idxs]),
        np.sum(X_avg[beta_idxs]),
        np.sum(X_avg[gamma_idxs])
    ]
    
    # Determina qual banda é a dominante (maior potência)
    dominant_band = np.argmax(bands)
    
    # Identificando se a banda dominante nesse trecho é beta ou gamma
    if dominant_band == 2 or dominant_band == 3:
        samples['focus'].append(X_avg) # se sim, rotula esse trecho como "focus"
    else:
        samples['not_focus'].append(X_avg) # se não, rotula como "not_focus"
        
    # Adicionando o rótulo correspondente
    # labels.append(band_labels.get(dominant_band))
    
    return samples

### Janela Deslizante

In [23]:

sr = 250            # Taxa de amostragem dos dados do EEG (250Hz)
jump = 5            # Tamanho da janela de análise (5 em 5 segundos)
size = sr * jump    # Número de pontos de dados em cada janela de análise

# Definindo os limites das bandas de frequência (em Hz)
theta_band = (4, 8)       # Theta: 4 - 8 Hz
alpha_band = (8, 13)      # Alpha: 8 - 13 Hz
beta_band = (13, 30)      # Beta: 13 - 30 Hz
gamma_band = (30, 100)    # Gamma: 30 - 100 Hz

# X dados, y label
samples = {
    'X': [],
    'Y': []
}

data_names = ('ia',)

# Mapeamento de índices para rótulos
band_labels = {
    0: 'focus',
    1: 'not_focus',
}

for k, data in enumerate(X.values()): # TODO: obter apenas os dados da aula e não do basal
    labels = [] # Lista para armazenar os rótulos das janelas

    data.set_eeg_reference(ref_channels='average', projection=False)
    
    # Percorre os dados em incrementos de sr (1 segundo) extraindo janelas de tamanho size (5s)
    for i in range(0, int(data.times[-1] - jump), 1):
        tmin = i
        tmax = i + jump

        cut = data.copy().crop(tmin=tmin, tmax=tmax).get_data()

        # Número de pontos por segmento
        n_per_seg = 128
        # Quantidade de sobreposição entre segmentos
        n_overlap = n_per_seg // 2 # 64
        
        # Calculando a densidade espectral de potência (PSD)
        freqs, cut_psd = welch(cut, fs=sr, nperseg=n_per_seg, noverlap=n_overlap)
        
        # Rotulando a amostra de 5 seg (cut_psd) como "focus" ou "not_focus"
        samples = label_data(freqs, cut_psd, samples)


print(samples)

EEG channel type selected for re-referencing
Applying average reference.
Applying a custom ('EEG',) reference.


EEG channel type selected for re-referencing
Applying average reference.
Applying a custom ('EEG',) reference.
{'focus': [array([5.44260454e-01, 6.23151676e+00, 1.83943684e+01, 1.07147014e+01,
       3.70877098e+00, 2.28004818e+00, 1.82455026e+00, 1.40667819e+00,
       9.43457589e-01, 7.31678030e-01, 6.69403179e-01, 7.49755134e-01,
       7.23110907e-01, 6.04842964e-01, 5.23307907e-01, 6.66435387e-01,
       1.06777404e+00, 1.05952147e+00, 8.78282090e-01, 6.18384837e-01,
       1.19085895e+00, 1.08671285e+00, 5.92924821e-01, 4.83595245e-01,
       6.96557551e-01, 7.26855806e-01, 6.43025326e-01, 8.61018956e-01,
       6.97477735e-01, 5.88576161e-01, 7.72250938e-01, 8.51006042e-01,
       1.25540585e+00, 1.03663040e+00, 7.91051925e-01, 7.80666940e-01,
       7.08931529e-01, 5.24980165e-01, 7.15698527e-01, 6.08588985e-01,
       6.17524603e-01, 7.67106982e-01, 5.86413592e-01, 5.18466172e-01,
       5.64363375e-01, 4.71847142e-01, 3.73510850e-01, 7.15416797e-01,
       4.03611997e-01, 4.1

### Estimando o ruído de fundo através do sinal basal

In [24]:
# Convertendo o tipo dos dados filtrados de objeto RawArray para ndarray
ndarray_data_basal = X['basal'].copy().get_data()

# Transformando os dados de basal para o domínio da frequência
_, data_basal_frequency = welch(ndarray_data_basal, fs=250, nperseg=512, noverlap=256)

# Lista que armazena as médias de potência para cada canal
noise_power = []

for channel_data in data_basal_frequency:
    # fft_result = np.fft.fft(channel_data)

    # # densidade espectral de potência (PSD)
    # psd = np.abs(fft_result) ** 2
    
    # média da potência no intervalo de tempo sem estímulo
    base_power = np.mean(channel_data)
    noise_power.append(base_power)

# média das médias de potência de todos os canais para estimar o ruído de fundo
estimated_background_noise = np.mean(noise_power)
print(estimated_background_noise)

0.4084735836541534


### Relação Sinal-Ruído de Banda Estreita

Banda estreita refere-se a sinais em que a largura de banda do sinal é pequena em relação à frequência central do sinal. Nesse sentido, a SNR de banda estreita é aplicada quando o sinal e o ruído ocupam uma faixa de frequência muito restrita. Ela pode ser observada pela seguinte equação:

$SNR_{banda\ estreita} = 10 \cdot \log_{10}\left(\frac{\text{energia total do espectro}}{\text{média das amplitudes nas frequências vizinhas}}\right)$

### Relação Sinal-Ruído de Banda Larga

Por outro lado, banda larga refere-se a sinais onde a largura de banda do sinal é grande, abrangendo uma ampla faixa de frequências. Dessa maneira, a SNR de banda larga aplica-se a situações nas quais tanto o sinal quanto o ruído ocupam uma larga faixa de frequências é definido da seguinte forma:

$SNR_{banda\ larga} = 10 \cdot \log_{10}\left(\frac{\text{energia total do espectro}}{\text{energia total do espectro de amplitude}}\right)$

### Aplicando as SNR de Banda Estreita e Larga para o sinal de interesse

In [25]:
# Lista de ndarrays (aqui, cada ndarray é uma amostra de 5 seg de dados de foco)
array_of_ndarrays = samples['focus']
print("array_of_ndarrays:")
print(array_of_ndarrays)
print(type(array_of_ndarrays))

narrow_band_SNR_list = []
wide_band_SNR_list = []

for ndarray in array_of_ndarrays:
    # Subtração matricial entre cada elemento de uma lista de amostra e o ruído de fundo (sinal basal)
    target_amplitudes_adjusted = ndarray - estimated_background_noise

    # Obtendo a SNR de banda estreita
    narrow_band_SNR = 10 * np.log10(target_amplitudes_adjusted / estimated_background_noise)
    print(narrow_band_SNR)
    narrow_band_SNR_list.append(narrow_band_SNR)
    print(narrow_band_SNR.shape)

    # Obtendo a SNR de banda larga
    total_power = np.sum(target_amplitudes_adjusted)
    wide_band_SNR = 10 * np.log10(target_amplitudes_adjusted / total_power)
    print(wide_band_SNR)
    wide_band_SNR_list.append(wide_band_SNR)
    print(wide_band_SNR.shape)

array_of_ndarrays:


[array([5.44260454e-01, 6.23151676e+00, 1.83943684e+01, 1.07147014e+01,
       3.70877098e+00, 2.28004818e+00, 1.82455026e+00, 1.40667819e+00,
       9.43457589e-01, 7.31678030e-01, 6.69403179e-01, 7.49755134e-01,
       7.23110907e-01, 6.04842964e-01, 5.23307907e-01, 6.66435387e-01,
       1.06777404e+00, 1.05952147e+00, 8.78282090e-01, 6.18384837e-01,
       1.19085895e+00, 1.08671285e+00, 5.92924821e-01, 4.83595245e-01,
       6.96557551e-01, 7.26855806e-01, 6.43025326e-01, 8.61018956e-01,
       6.97477735e-01, 5.88576161e-01, 7.72250938e-01, 8.51006042e-01,
       1.25540585e+00, 1.03663040e+00, 7.91051925e-01, 7.80666940e-01,
       7.08931529e-01, 5.24980165e-01, 7.15698527e-01, 6.08588985e-01,
       6.17524603e-01, 7.67106982e-01, 5.86413592e-01, 5.18466172e-01,
       5.64363375e-01, 4.71847142e-01, 3.73510850e-01, 7.15416797e-01,
       4.03611997e-01, 4.14984323e-01, 3.30338187e-01, 2.14696436e-01,
       3.53711964e-01, 2.33014094e-01, 1.96499688e-01, 2.11698432e-01,
     

/tmp/ipykernel_3823/1509557533.py:15: RuntimeWarning: invalid value encountered in log10
  narrow_band_SNR = 10 * np.log10(target_amplitudes_adjusted / estimated_background_noise)
/tmp/ipykernel_3823/1509557533.py:22: RuntimeWarning: invalid value encountered in log10
  wide_band_SNR = 10 * np.log10(target_amplitudes_adjusted / total_power)


[         nan -18.46222585 -12.70693993 -13.98407635 -17.19067451
 -20.98486237 -22.22575083 -22.95173518 -21.85165242 -22.6929985
 -20.08341885 -21.53896277 -20.67107075 -15.62712381 -15.00886033
 -15.4139311  -15.85594155 -15.83360807 -15.50503989 -15.60191835
 -14.78300397 -14.38681949 -15.61443321 -17.94617114 -16.30218171
 -16.01125808 -15.50383747 -15.51089199 -17.50486348 -17.38670659
 -17.33493845 -16.61246288 -14.70641269 -15.35682422 -15.19954074
 -16.64781421 -15.17226705 -14.41367817 -17.88898624 -19.85986945
 -19.17618367 -18.54058234 -17.51124498 -18.64141934 -17.77143145
 -18.15415084 -19.3263351  -21.10878132 -20.70913613 -21.83177553
 -19.83226374 -21.62787485 -24.19935536 -22.257627   -22.03330377
          nan          nan          nan          nan          nan
          nan          nan          nan          nan          nan]
(65,)
[        nan  4.39680103 10.13791624  8.87781472  4.99691991  0.55035765
  0.44361271  0.40984507  1.83176263  1.21235154  3.44402683  3

In [28]:
print(len(wide_band_SNR_list))
len(wide_band_SNR)

162


65

### Tarefa para aplicação das características SNR:

Agora que temos os dois vetores de características SNR, podemos utilizar buffers com e sem a evocação dos rítmos que caracterizam o foco.

#### Divisão dos dados

Utilizando a iteração (por exemplo, de 5 segundos caracterizada pela janela) realizada no sinal a cada ~1 segundo, realize a rotulação dos dados de interesse (Beta e Gamma). Ou seja, cada amostra será um sinal de 5 segundos (1250 pontos de 8 canais). A janela que não for qualificada como Beta ou Gama por exemplo, poderá ser rotulada com "desfoque". Se acharem interessante, adicionar rótulos do ritmo Theta também.

No caso do sinal que representa o basal (se tiverem) poderá pegar um único sinal de aproximadamente 30 segundos para ser utilizado na equação de ruído, que irá ter como resultado um único valor. Lembrando que o valor de ruído deve atuar no sinal no domínio da frequência.

#### Classificação

Cálculo: diferença entre tempo inicial (em minutos) e final (tbm em min) dada por: (tempo_min * 60) * 250. Aplicando aos dados reais para criar o vetor de características:

- Diferença: 90000 - 154250 com 8 canais
- Pontos totais: 64250 com 8 canais

- 64.250 (pontos totais) / 250 (taxa de amostragem) = 257 segundos
- 257 / 5 (tamanho da janela sem sobreposição) = 51 amostras (51,4)

<br>

Em nossos dados simulados, temos 150.000 pontos com 8 canais. A utilização desses dados funcionará da seguinte forma para a criação do vetor de características:

- 150.000 (pontos totais) / 250 (taxa de amostragem) = 600 segundos
- 600 / 5 (tamanho da janela sem sobreposição) = 120 amostras

| 1   | SNRw1                | SNRw2 | ... | SNRw8 | SNRn1 | SNRn2 | ... | SNRn8 |
|-----|----------------------|-------|-----|-------|-------|-------|-----|-------|
| 2   | [w1, w2, ..., w1250] |       |     |       |       |       |     |       |
| 3   |                      |       |     |       |       |       |     |       |
| ... |                      |       |     |       |       |       |     |       |
| 120 |                      |       |     |       |       |       |     |       |

- Agora transforme cada um dos vetores de pontos no domínio da frequência (1250 pontos) em um único valor real. Neste caso pode ser utilizado tanto a média como a mediana (ou ambos). Se utilizarmos as duas, teremos no final 32 colunas de características:
    - 8 canais
    - SNR narrow e SNR wide (2)
    - Média e mediana (2)

| 1   | 1   | ... | 32 |
|-----|-----|-----|----|
| 2   | w'  | ... |    |
| 3   | ... |     |    |
| ... |     |     |    |
| 120 |     |     |    |



Após obter o vetor de característica, realizar a divisão dos dados em treinamento e teste (normalmente uma proporção de 70 e 30% respectivamente) e aplicar para o classificador SVM.

**PLUS**: Ao final da tarefa, verificar a melhora dos resultados utilizando um seletor de características. Neste caso, podemos utilizar o RFE (*Recursive Feature Elimination*) em uma fase anterior a classificação para reduzir as 32 características se for necessário.


